Allow hiring of employed workers

mcmanustfj opened the pull request:
See [old PR](https://github.com/ldtteam/minecolonies/pull/7193) for some context.  In short, this PR makes it so you don't have to fire a citizen before hiring them somewhere else.  This was rejected a year ago with the feedback that I should add a toggle, so I did.

# Changes proposed in this pull request:

- This feature will allow huts hiring for "normal" (`WorkerBuildingModuleView`) jobs to hire workers that are already employed, firing them from their current job.  Special huts (quarry, warehouse, school hiring students) will not be able to do this.* 
- Workers with jobs are listed after unemployed workers and workers who would normally be able to be hired (i.e. from the library).
- This feature is disabled by default, with a button in the hire window screen to allow it.  This button is disabled for huts that can't use this feature.


This is what the UI looks like when showing employed:
![image](https://user-images.githubusercontent.com/10820772/157578859-aacf28c4-0c68-4c0b-be67-015c8175e868.png)

## Improvements that should be made before this is ready
 - Right now, there's supposed to be a Y or N in the 'show employed' button, but it's not showing up and I don't know why.  Please take a look and advise, because I can't figure this out.
 - ~~The button is still appearing for pupils when it shouldn't be (and lets you hire adults as pupils).~~ (fixed)
 - I moved the "Turn the hiring mode of this hut to manual to fire this worker" warning above the building hiring mode, which slightly infringes on the worker list view.  I think this is fine in most circumstances (because you wouldn't be doing this on automatic hiring mode) but it's still a slight eyesore.  I don't know how to do this better, so I'd love feedback.
![image](https://user-images.githubusercontent.com/10820772/157579373-9551e111-1f05-4228-9e89-66168dcf2fa3.png)

Review please

mcmanustfj committed:
Made EasyHire work in 1.18

mcmanustfj committed:
Added toggle for hiring employed workers

mcmanustfj committed:
Fixed pupil issue, probably

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java:
```diff
@@ -207,45 +221,89 @@ private void doneClicked(@NotNull final Button button)
         final int row = citizenList.getListElementIndexByPane(button);
         @NotNull final ICitizenDataView citizen = citizens.get(row);
 
```
Formatting is off. we do:

{

}

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java:
```diff
@@ -207,45 +221,89 @@ private void doneClicked(@NotNull final Button button)
         jobList.refreshElementPanes();
     }
 
+    protected void showEmployedToggled(@NotNull final Button button)
```
javadoc

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java:
```diff
@@ -207,45 +221,89 @@ private void doneClicked(@NotNull final Button button)
+        onOpened();
+    }
+
+    private void setupShowEmployed()
```
javadoc

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java:
```diff
@@ -207,45 +221,89 @@ private void doneClicked(@NotNull final Button button)
+        showEmployed = false;
+    }
+
+    private boolean canAssign(ICitizenDataView citizen)
```
javadoc

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHireWorker.java:
```diff
@@ -207,45 +221,89 @@ private void doneClicked(@NotNull final Button button)
+        // Fire citizen if they already have a job
+        if (citizen.getWorkBuilding() != null && selectedModule instanceof WorkerBuildingModuleView) {
+            IBuildingView oldJob =  colony.getBuilding(citizen.getWorkBuilding());
+            oldJob.getModuleViewMatching(IAssignmentModuleView.class,
```
Does this trigger it on th eserverside too?

mcmanustfj replied:
It should? It uses the same `.removeCitizen` function that `fireClicked()` does.

Raycoms replied:
kk

Raycoms reviewed (changes requested):
Several comments

mcmanustfj committed:
Feedback and researcher XP nerfs

mcmanustfj committed:
Merge branch 'version/1.18' of https://github.com/ldtteam/minecolonies into easyhire118

mcmanustfj commented:
Added the researcher XP nerfs as discussed in discord.  Picked 5 out of a hat as an XP number - should be the same XP for the primary stat (which was 0.7) at like 14 int (without housing/building modifiers) but totally removed the absurd mana xp they were getting

mcmanustfj committed:
Accidentally gave way too much XP

mcmanustfj commented:
~~whoops was giving 5 xp instead of .5 xp~~
giving 5 xp before int reductions was intentional, but when playtesting it was still way too high - gained like 7 levels in one day with 30 int and a 60% building bonus.  Will bump back up a bit.

mcmanustfj committed:
Increased XP a bit

mcmanustfj committed:
Balanced xp according to discord conversation

so for researcher since the work day is from time 0 to time 10600 (let's
say 60 seconds on each end for commute, so 8200 working ticks)  and they
spend it walking to a bookshelf (depends on schematic, we'll say 10
seconds/200 ticks) and opening a book for 60 ticks, they probably get
experience 32 times a day.
10/32 -> 0.3125 xp per study at int 1 -> 3.125 xp given

someaddons commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/research/EntityAIWorkResearcher.java:
```diff
@@ -23,9 +23,9 @@
+     * base XP gained per study position
      */
-    private static final double XP_PER_STUDYPOS = 0.7;
+    private static final double XP_PER_STUDYPOS = 3.125;
```
I think XP and studies/day are still off, currently we have a study delay of just 60 ticks(3seconds) + walking time inbetween, which is likely more than just ~3 studies a day you need for getting to the average 10xp base.

A minecraft day is 10 minutes, ~3 of those are spent at getting to or from work, so lets say 7mins effective working time. If they now study for 3s + ~20s walking time you get a total of 18 study actions a day.
Following the baseline of 10Xp a day a worker should get, you'll end up giving them 10/18 = ~0.55 xp per study action.

I think we should do two things here: 
First increase the study delay, 60 ticks is very fast I think they should be looking at a "book" near a bookcase for about 60 seconds instead which would be 60 * 20 ticks.

When we study every 60 seconds and estimate another 20seconds of walking time we get 80 seconds per study, with the estimated minecraft work day of 7minutes that is ~ 5 Study actions per day. 

At 10 Xp Base and 5 study actions a day you'd give it 2 XP per study action.

mcmanustfj replied:
Will do!  I misread the XP calculation, I thought that

`localXp += localXp * (intelligenceLevel / 100.0);`

was actually 

`localXp = localXp * (intelligenceLevel / 100.0);`

(in which case I should have given them 31.25 xp per action).  
I'll do the things you suggested!

mcmanustfj committed:
Fixed researcher XP for the last time
